File: src/main/java/sf/llk/share/util/ISpaceUtil.java
/*!            
    C+edition for Desktop, Community Edition.
    Copyright (C) 2021 Cplusedition Limited.  All rights reserved.
    
    The author licenses this file to You under the Apache License, Version 2.0
    (the "License"); you may not use this file except in compliance with
    the License.  You may obtain a copy of the License at
    
        http://www.apache.org/licenses/LICENSE-2.0
    
    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
*/
/*
 * The software in this package is distributed under the GNU General Public
 * License version 2, as published by the Free Software Foundation, but with
 * the Classpath exception.  You should have received a copy of the GNU General
 * Public License (GPL) and the Classpath exception along with this program.
 */
package sf.llk.share.util;

import java.util.Collection;
import java.util.List;

public interface ISpaceUtil extends ISpaceDetector {

    /**
     * @return true if character is a space character.
     */
    @Override
    public abstract boolean isSpace(char c);

    /**
     * @return true if character is a line break character.
     */
    public abstract boolean isLineBreak(char c);

    /**
     * @return true if character is a space or line break character.
     */
    @Override
    public abstract boolean isWhitespace(char c);

    public abstract boolean isSpaces(CharSequence s);

    public abstract boolean isSpaces(CharSequence s, int start, int end);

    public abstract boolean isWhitespaces(CharSequence s);

    public abstract boolean isWhitespaces(CharSequence s, int start, int end);

    @Override
    public abstract int skipSpaces(CharSequence s, int start, int end);

    @Override
    public abstract int skipWhitespaces(CharSequence s, int start, int end);

    @Override
    public abstract int rskipSpaces(CharSequence s, int start, int end);

    @Override
    public abstract int rskipWhitespaces(CharSequence s, int start, int end);

    public abstract int skipNonSpaces(CharSequence s, int start, int end);

    public abstract int skipNonWhitespaces(CharSequence s, int start, int end);

    public abstract int rskipNonSpaces(CharSequence s, int start, int end);

    public abstract int rskipNonWhitespaces(CharSequence s, int start, int end);

    public abstract boolean endsWithLineBreak(CharSequence s);

    public abstract boolean endsWithSpaces(CharSequence s);

    public abstract boolean endsWithWhitespaces(CharSequence s);

    public abstract int countLines(CharSequence s);

    public abstract int countLines(CharSequence s, int start, int end);

    public abstract int lcountLineBreaks(CharSequence s);

    public abstract int lcountLineBreaks(CharSequence s, int start, int end);

    public abstract int lcountBlankLines(CharSequence s);

    public abstract int lcountBlankLines(CharSequence s, int start, int end);

    /**
     * @return Number of trailing line breaks.
     */
    public abstract int rcountLineBreaks(CharSequence s);

    public abstract int rcountLineBreaks(CharSequence s, int start, int end);

    /**
     * @return Number of trailing blank lines.
     */
    public abstract int rcountBlankLines(CharSequence s);

    public abstract int rcountBlankLines(CharSequence s, int start, int end);

    public abstract boolean hasLineBreak(CharSequence s);

    public abstract boolean hasLineBreak(CharSequence s, int start, int end);

    public abstract boolean hasWhitespace(CharSequence s, int start, int end);

    public abstract CharSequence ltrimSpaces(CharSequence b);

    public abstract CharSequence ltrimSpaces(CharSequence b, int start, int end);

    public abstract CharSequence ltrimWhitespaces(CharSequence b);

    public abstract CharSequence ltrimWhitespaces(CharSequence b, int start, int end);

    public abstract CharSequence ltrimLineBreaks(CharSequence buf);

    public abstract CharSequence ltrimBlankLines(CharSequence buf);

    public abstract CharSequence rtrimSpaces(CharSequence b);

    public abstract CharSequence rtrimSpaces(CharSequence b, int start, int end);

    public abstract CharSequence rtrimWhitespaces(CharSequence b);

    public abstract CharSequence rtrimWhitespaces(CharSequence b, int start, int end);

    public abstract CharSequence rtrimLineBreaks(CharSequence buf);

    public abstract CharSequence rtrimBlankLines(CharSequence buf);

    public abstract CharSequence trimSpaces(CharSequence b);

    public abstract CharSequence trimSpaces(CharSequence b, int start, int end);

    public abstract CharSequence trimWhitespaces(CharSequence b);

    public abstract CharSequence trimWhitespaces(CharSequence b, int start, int end);

    public abstract boolean ltrimSpaces(StringBuilder b);

    public abstract boolean ltrimSpaces(StringBuilder b, int start, int end);

    public abstract boolean ltrimWhitespaces(StringBuilder b);

    public abstract boolean ltrimWhitespaces(StringBuilder b, int start, int end);

    /**
     * @return Number of trailing line breaks deleted.
     */
    public abstract int ltrimLineBreaks(StringBuilder buf);

    /**
     * @return Number of blank lines deleted.
     */
    public abstract int ltrimBlankLines(StringBuilder buf);

    public abstract boolean rtrimSpaces(StringBuilder b);

    public abstract boolean rtrimSpaces(StringBuilder b, int start, int end);

    public abstract boolean rtrimWhitespaces(StringBuilder b);

    public abstract boolean rtrimWhitespaces(StringBuilder b, int start, int end);

    /**
     * @return Number of trailing line breaks deleted.
     */
    public abstract int rtrimLineBreaks(StringBuilder buf);

    /**
     * @return Number of blank lines deleted.
     */
    public abstract int rtrimBlankLines(StringBuilder buf);

    public abstract boolean rspace(StringBuilder buf);

    public abstract boolean ltrimSpaces(StringBuffer b);

    public abstract boolean ltrimSpaces(StringBuffer b, int start, int end);

    public abstract boolean ltrimWhitespaces(StringBuffer b);

    public abstract boolean ltrimWhitespaces(StringBuffer b, int start, int end);

    /**
     * @return Number of trailing line breaks deleted.
     */
    public abstract int ltrimLineBreaks(StringBuffer buf);

    /**
     * @return Number of blank lines deleted.
     */
    public abstract int ltrimBlankLines(StringBuffer buf);

    public abstract boolean rtrimSpaces(StringBuffer b);

    public abstract boolean rtrimSpaces(StringBuffer b, int start, int end);

    public abstract boolean rtrimWhitespaces(StringBuffer b);

    public abstract boolean rtrimWhitespaces(StringBuffer b, int start, int end);

    /**
     * @return Number of trailing line breaks deleted.
     */
    public abstract int rtrimLineBreaks(StringBuffer buf);

    /**
     * @return Number of blank lines deleted.
     */
    public abstract int rtrimBlankLines(StringBuffer buf);

    public abstract boolean rspace(StringBuffer buf);

    public abstract int columnOf(CharSequence s, int start, int end, int lmargin, int tabwidth);

    public abstract List<String> splitLines(CharSequence str);

    /**
     * Split words using whitespaces as delimiters.
     */
    public abstract int splitWords(Collection<String> ret, CharSequence str, int start, int end);
}

File: src/main/java/com/cplusedition/bot/text/ISpaceUtil.kt
/*!            
    C+edition for Desktop, Community Edition.
    Copyright (C) 2021 Cplusedition Limited.  All rights reserved.
    
    The author licenses this file to You under the Apache License, Version 2.0
    (the "License"); you may not use this file except in compliance with
    the License.  You may obtain a copy of the License at
    
        http://www.apache.org/licenses/LICENSE-2.0
    
    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
*/
package com.cplusedition.bot.text

interface ISpaceUtil : ISpaceDetector {

    /** @return true if character is a line break character.
     */
    fun isLineBreak(c: Char): Boolean

    fun isSpaces(s: CharSequence): Boolean
    fun isSpaces(s: CharSequence, start: Int, end: Int): Boolean
    fun isWhitespaces(s: CharSequence): Boolean
    fun isWhitespaces(s: CharSequence, start: Int, end: Int): Boolean

    fun skipNonSpaces(s: CharSequence, start: Int, end: Int): Int
    fun skipNonWhitespaces(s: CharSequence, start: Int, end: Int): Int
    fun rskipNonSpaces(s: CharSequence, start: Int, end: Int): Int
    fun rskipNonWhitespaces(s: CharSequence, start: Int, end: Int): Int

    fun endsWithLineBreak(s: CharSequence): Boolean
    fun endsWithSpaces(s: CharSequence): Boolean
    fun endsWithWhitespaces(s: CharSequence): Boolean

    fun countLines(s: CharSequence): Int
    fun countLines(s: CharSequence, start: Int, end: Int): Int

    fun lcountLineBreaks(s: CharSequence): Int
    fun lcountLineBreaks(s: CharSequence, start: Int, end: Int): Int
    fun lcountBlankLines(s: CharSequence): Int
    fun lcountBlankLines(s: CharSequence, start: Int, end: Int): Int

    /** @return Number of trailing line breaks.
     */
    fun rcountLineBreaks(s: CharSequence): Int

    fun rcountLineBreaks(s: CharSequence, start: Int, end: Int): Int
    /** @return Number of trailing blank lines.
     */
    fun rcountBlankLines(s: CharSequence): Int

    fun rcountBlankLines(s: CharSequence, start: Int, end: Int): Int

    fun hasLineBreak(s: CharSequence): Boolean
    fun hasLineBreak(s: CharSequence, start: Int, end: Int): Boolean
    fun hasWhitespace(s: CharSequence, start: Int, end: Int): Boolean

    fun ltrimSpaces(b: CharSequence): CharSequence
    fun ltrimSpaces(b: CharSequence, start: Int, end: Int): CharSequence
    fun ltrimWhitespaces(b: CharSequence): CharSequence
    fun ltrimWhitespaces(b: CharSequence, start: Int, end: Int): CharSequence
    fun ltrimLineBreaks(buf: CharSequence): CharSequence
    fun ltrimBlankLines(buf: CharSequence): CharSequence

    fun rtrimSpaces(b: CharSequence): CharSequence
    fun rtrimSpaces(b: CharSequence, start: Int, end: Int): CharSequence
    fun rtrimWhitespaces(b: CharSequence): CharSequence
    fun rtrimWhitespaces(b: CharSequence, start: Int, end: Int): CharSequence
    fun rtrimLineBreaks(buf: CharSequence): CharSequence
    fun rtrimBlankLines(buf: CharSequence): CharSequence

    fun trimSpaces(b: CharSequence): CharSequence
    fun trimSpaces(b: CharSequence, start: Int, end: Int): CharSequence
    fun trimWhitespaces(b: CharSequence): CharSequence
    fun trimWhitespaces(b: CharSequence, start: Int, end: Int): CharSequence

    fun ltrimSpaces(b: StringBuilder): Boolean
    fun ltrimSpaces(b: StringBuilder, start: Int, end: Int): Boolean
    fun ltrimWhitespaces(b: StringBuilder): Boolean
    fun ltrimWhitespaces(b: StringBuilder, start: Int, end: Int): Boolean
    /** @return Number of trailing line breaks deleted.
     */
    fun ltrimLineBreaks(buf: StringBuilder): Int

    /** @return Number of blank lines deleted.
     */
    fun ltrimBlankLines(buf: StringBuilder): Int

    fun rtrimSpaces(b: StringBuilder): Boolean
    fun rtrimSpaces(b: StringBuilder, start: Int, end: Int): Boolean
    fun rtrimWhitespaces(b: StringBuilder): Boolean
    fun rtrimWhitespaces(b: StringBuilder, start: Int, end: Int): Boolean
    /** @return Number of trailing line breaks deleted.
     */
    fun rtrimLineBreaks(buf: StringBuilder): Int

    /** @return Number of blank lines deleted.
     */
    fun rtrimBlankLines(buf: StringBuilder): Int

    fun rspace(buf: StringBuilder): Boolean

    fun ltrimSpaces(b: StringBuffer): Boolean
    fun ltrimSpaces(b: StringBuffer, start: Int, end: Int): Boolean
    fun ltrimWhitespaces(b: StringBuffer): Boolean
    fun ltrimWhitespaces(b: StringBuffer, start: Int, end: Int): Boolean
    /** @return Number of trailing line breaks deleted.
     */
    fun ltrimLineBreaks(buf: StringBuffer): Int

    /** @return Number of blank lines deleted.
     */
    fun ltrimBlankLines(buf: StringBuffer): Int

    fun rtrimSpaces(b: StringBuffer): Boolean
    fun rtrimSpaces(b: StringBuffer, start: Int, end: Int): Boolean
    fun rtrimWhitespaces(b: StringBuffer): Boolean
    fun rtrimWhitespaces(b: StringBuffer, start: Int, end: Int): Boolean
    /** @return Number of trailing line breaks deleted.
     */
    fun rtrimLineBreaks(buf: StringBuffer): Int

    /** @return Number of blank lines deleted.
     */
    fun rtrimBlankLines(buf: StringBuffer): Int

    fun rspace(buf: StringBuffer): Boolean

    fun columnOf(s: CharSequence, start: Int, end: Int, lmargin: Int, tabwidth: Int): Int
    fun splitLines(str: CharSequence): List<String>
    /** Split words using whitespaces as delimiters.  */
    fun splitWords(ret: MutableCollection<String>, str: CharSequence, start: Int, end: Int): Int
}
